target/ppc/translate/vsx-impl.c.inc

```diff
@@ -2124,6 +2124,247 @@ static void gen_xvxsigdp(DisasContext *ctx)
+    tcg_gen_andi_i64(cond, xc, 0x0000000080000000);
+    tcg_gen_movcond_i64(TCG_COND_EQ, t1, zero, cond, xa, xb);
+    tcg_gen_andi_i64(t1, t1, 0x00000000ffff0000);
+    tcg_gen_or_i64(t0, t0, t1);
+    tcg_gen_andi_i64(cond, xc, 0x0000000000008000);
+    tcg_gen_movcond_i64(TCG_COND_EQ, t1, zero, cond, xa, xb);
+    tcg_gen_andi_i64(t1, t1, 0x000000000000ffff);
+    tcg_gen_or_i64(t0, t0, t1);
+    set_cpu_vsrl(a->xt, t0);
+
+    tcg_temp_free_i64(xc);
+    tcg_temp_free_i64(xb);
+    tcg_temp_free_i64(xa);
+    tcg_temp_free_i64(t0);
+    tcg_temp_free_i64(t1);
+    tcg_temp_free_i64(cond);
+    return true;
+}
+
+static bool trans_XXBLENDVW(DisasContext *ctx, arg_XX4 *a)
+{
+    TCGv_i64 t0, t1, cond, zero;
+    TCGv_i64 xa, xb, xc;
+    REQUIRE_VSX(ctx);
+
+    t0 = tcg_temp_new_i64();
+    t1 = tcg_temp_new_i64();
+    cond = tcg_temp_new_i64();
+    xa = tcg_temp_new_i64();
+    xb = tcg_temp_new_i64();
+    xc = tcg_temp_new_i64();
+    zero = tcg_constant_i64(0);
+
+    get_cpu_vsrh(xa, a->xa);
+    get_cpu_vsrh(xb, a->xb);
+    get_cpu_vsrh(xc, a->xc);
+    tcg_gen_andi_i64(cond, xc, 0x8000000000000000);
+    tcg_gen_movcond_i64(TCG_COND_EQ, t0, zero, cond, xa, xb);
+    tcg_gen_andi_i64(t0, t0, 0xffffffff00000000);
+    tcg_gen_andi_i64(cond, xc, 0x0000000080000000);
+    tcg_gen_movcond_i64(TCG_COND_EQ, t1, zero, cond, xa, xb);
+    tcg_gen_andi_i64(t1, t1, 0x00000000ffffffff);
+    tcg_gen_or_i64(t0, t0, t1);
+    set_cpu_vsrh(a->xt, t0);
+
+    get_cpu_vsrl(xa, a->xa);
+    get_cpu_vsrl(xb, a->xb);
+    get_cpu_vsrl(xc, a->xc);
+    tcg_gen_andi_i64(cond, xc, 0x8000000000000000);
+    tcg_gen_movcond_i64(TCG_COND_EQ, t0, zero, cond, xa, xb);
+    tcg_gen_andi_i64(t0, t0, 0xffffffff00000000);
+    tcg_gen_andi_i64(cond, xc, 0x0000000080000000);
+    tcg_gen_movcond_i64(TCG_COND_EQ, t1, zero, cond, xa, xb);
+    tcg_gen_andi_i64(t1, t1, 0x00000000ffffffff);
+    tcg_gen_or_i64(t0, t0, t1);
+    set_cpu_vsrl(a->xt, t0);
+
+    tcg_temp_free_i64(xc);
+    tcg_temp_free_i64(xb);
+    tcg_temp_free_i64(xa);
+    tcg_temp_free_i64(t0);
+    tcg_temp_free_i64(t1);
+    tcg_temp_free_i64(cond);
+    return true;
+}
+
+static bool trans_XXBLENDVD(DisasContext *ctx, arg_XX4 *a)
+{
+    TCGv_i64 t0;
+    TCGv_i64 xc, xa, xb;
+    TCGv_i64 zero;
+
+    REQUIRE_VSX(ctx);
+
+    t0 = tcg_temp_new_i64();
+    xa = tcg_temp_new_i64();
+    xb = tcg_temp_new_i64();
+    xc = tcg_temp_new_i64();
+    zero = tcg_constant_i64(0);
+
+    /* first double-word */
+    get_cpu_vsrh(xa, a->xa);
+    get_cpu_vsrh(xb, a->xb);
+    get_cpu_vsrh(xc, a->xc);
+    tcg_gen_andi_i64(xc, xc, 0x8000000000000000);
+    tcg_gen_movcond_i64(TCG_COND_EQ, t0, zero, xc, xa, xb);
+    set_cpu_vsrh(a->xt, t0);
+    /* second double-word */
+    get_cpu_vsrl(xa, a->xa);
+    get_cpu_vsrl(xb, a->xb);
+    get_cpu_vsrl(xc, a->xc);
+    tcg_gen_andi_i64(xc, xc, 0x8000000000000000);
+    tcg_gen_movcond_i64(TCG_COND_EQ, t0, zero, xc, xa, xb);
+    set_cpu_vsrl(a->xt, t0);
```

Comment:
We could use a unified function that all BLEND instructions call, saving some lines, but we'd have to see the impact on performance. But that can also be for a later clean up.

Reply:
Just to have it here and not lose the data, I have implemented a helper for the blendvb function (that could've been turned into a macro and dealt with other blends) and because it was writing in global variables, it turned out to be about 6x slower.

If at some point later it was decided to try and implement one without using direct pointers to VSR, instead getting and setting them in the trans_BLEND function, it may not have such a severe performance impact